### PR TITLE
Enable Server to Communicate with Multiple Clients: Enhance Client-Server Synchronization and Refactor Server Notification Handling

### PR DESCRIPTION
**Pull request by Amithabh-A**
#### Description
This pull request addresses the following issues:
- Only one client was able to connect to the server at a time, limiting concurrent communication capabilities.

#### Changes Made
- **Feature/Enhancement**:  
  - Added a `BinarySemaphore` class to manage synchronization between the server and multiple clients.  
  - Modified the client-server communication flow by introducing a `SyncUp` request.  
  - Clients now acquire the semaphore before making sync requests, ensuring orderly communication between server and clients. The server uses the `SyncUp` request to wait for client synchronization, and the client sends a `SyncUp` request after the UI triggers a sync.

- **Bug Fix**:  
  - Resolved the issue where only one client could connect to the server at a time. With the new synchronization mechanism, multiple clients can now communicate with the server in sequence.

- **Refactoring**:  
  - Renamed the class `ClientMetadataHandler` to `ServerNotificationHandler`, as the original name was misleading. The new name better reflects the server's role in handling notifications.  
  - Removed the nested `ServerNotificationHandler` class from within the `Server` class, as nested classes were deemed inappropriate in this case.

#### How to Test
1. Test the synchronization flow by starting a server and connecting multiple clients. Verify that each client can interact with the server in sequence by triggering the `SyncUp` request from the client UI

#### Additional Notes
- The introduction of the `BinarySemaphore` class ensures that the server does not handle multiple client requests concurrently but still allows for orderly, sequential communication.  
- If preferred, the `ServerNotificationHandler` class could be renamed to simply `NotificationHandler`, which would more closely align with its functionality. 

#### Checklist
- [ ] I have tested my changes locally. (not possible in mac when UI is there)
- [ ] I have added/updated tests to cover my changes. (later)
- [x] I have followed the coding standards of this project. (For the Server.cs)

### CODE REVIEW

File: Updater/Utils.cs [hunk 1]
```diff
@@ -85,7 +85,8 @@ public static string SerializeObject<T>(T obj)
             ISerializer serializer = new Serializer();
             return serializer.Serialize(obj);
         }
-        catch (Exception ex) { 
+        catch (Exception ex)
+        {
             Console.WriteLine(ex.ToString());
```
Comment (rg1013) [inline Updater/Utils.cs:90]: Don't use Console statements, use Trace

File: Updater/Server.cs [hunk 2]
```diff
@@ -35,312 +41,439 @@ public void Start(string ip, string port)
...
+    public static void UpdateUILogs(string message)
+    {
+        NotificationReceived?.Invoke(message);
+    }
+}
+
+/// <summary>
+/// INotificationHandler Implementation in Server side
+/// </summary>
+public class ServerNotificationHandler : INotificationHandler
```
Comment (rg1013) [inline Updater/Server.cs:142]: The notifier name should be same on both server and client side, pls check the networking code to understand better

File: Updater/Utils.cs (hunk 1 above)
Comment (Amithabh-A) [reply Updater/Utils.cs]: Lol I didn't do it
Anyways changed them

**Review by rg1013 — APPROVED** [review]

File: Updater/Client.cs [hunk 3]
```diff
@@ -44,17 +44,27 @@ public async Task<string> StartAsync(string ipAddress, string port)
...
         _communicator.Subscribe("ClientMetadataHandler", new ClientMetadataHandler(_communicator));
-        SyncUp();
+        // SyncUp();
     }
 
+
+    /// <summary>
+    /// Sends a SyncUp request to the server
+    /// </summary>
     public void SyncUp()
```
[the remark this answers is not in the document]
Comment (Amithabh-A) [reply Updater/Client.cs:54]: It is useful for sync button in the client ui.

**Pull request merged**